### PR TITLE
=== rather than == / array destructuring in place of slice

### PR DESCRIPTION
**Pull request by johnnyreilly**
Hey @graeme-lockley!

This is a PR that you can feel free to ignore!  But I thought it could be fun 😄 

I did this in the GitHub editor - I wish they'd hurry up and launch codespaces: https://github.com/features/codespaces

Nice use of GitHub Actions BTW!  It linted me the hard way. 

### Array destructuring in favour of `splice`

So there's some cools stuff you can do in JavaScript / TypeScript these days.  Where you would once have created a new array sans first element like so: 

```ts
docDocs.splice(1) //...
```

You can now do this using the ES6 destructuring gorgeousness! https://stackoverflow.com/a/59411548/761388

```ts
    const [theFirst, ...allButTheFirst] = docDocs;
    return allButTheFirst.reduce((a, b) => a.p(b), theFirst);
```

Which gives you a new array with all the elements but the first one, and puts the first one in a new variable to use.

It uses an extra line of code but it is for my money much clearer.  What do you think?

### Favouring `===`

I was wondering if you'd intended to use `==` instead of `===`?  You may not know about JavaScript's operator quirkiness, but the long story short is that if you'd use `!=` in another language you probably want `!==` in JS and where you'd use `==` in another language you probably want `===` in JS.  There's a better write up than I'd manage here:

https://stackoverflow.com/questions/359494/which-equals-operator-vs-should-be-used-in-javascript-comparisons

### Lovely stuff!

I really like your usage of `toDoc` directly with `.map` etc.  So pleasing to not need to introduce an anonymous function just for the purposes of calling on.

Also it's nice to see usage of `_` where you're ignoring a parameter in a function.  Completely subjective: I favour using a standalone variable rather than one wrapped by parens when there's only a single parameter in an arrow function.  Following the latest release of Prettier, the community is standardising on the approach you favour of having parens always:

```ts
// Input
const fn = (x) => y => x + y;

// Prettier 1.19
const fn = x => y => x + y;

// Prettier 2.0
const fn = (x) => (y) => x + y;
```

There's a good write up here: https://prettier.io/blog/2020/03/21/2.0.0.html#change-default-value-for-arrowparens-to-always-7430httpsgithubcomprettierprettierpull7430-by-kachkaevhttpsgithubcomkachkaev

It makes sense - I'm going to move to it (but with gritted teeth 😄 )

### CODE REVIEW

**Comment by graeme-lockley** [comment]
Thanks do much for taking the time and the pointers.

Yep === rather than == breaks my head.  I understand why bit it does seem like a crime against good visual design.

Sweet re the destructing of arrays - more elegant and I agree it reads easier.  The extra line is no problem at all because, for me at least, my eye flows over the code with understanding rather than having to read the code.

What is the general trend around TypeScript functions?  Should one only use functions when this is necessary otherwise use arrow functions?  My preference is to only use arrow functions but, on this code base, I have stuck with function and only use arrow functions for anonymous functions.  So even the exported function

```ts
export function nest(offset: number, doc: Doc | string): Doc {
  return new NestDoc(offset, toDoc(doc));
}
```

my preference is to write as

```ts
export const nest = (offset: number, doc: Doc | string): Doc =>
  new NestDoc(offset, toDoc(doc));
```

Whether or not to go full curried I'm not sure

```ts
export const nest = (offset: number) => (doc: Doc | string): Doc =>
  new NestDoc(offset, toDoc(doc));
```

The caller descends into parenthesis hell.  I tried that style for a while - the syntax and tooling doesn't support it well enough and the code has a completely different flavour.  

Me things my love for functional languages is too dominant and the resulting code is too far off idiomatic TypeScript.

Thanks again John for the comments!

**Comment by johnnyreilly** [comment]
#### Arrow functions vs classic functions

> What is the general trend around TypeScript functions? Should one only use functions when this is necessary otherwise use arrow functions?

Historically there was some issue around arrow functions not having great stack traces.  Having had a quick google I think that's a thing of the past though.  Essentially it's pretty much preference now.  There's a couple of factors here and there that might sway you.

- Arrow functions have a lexically bound `this` - this is great news in the context of `class`es but otherwise just incidental.
- There were issues with "find all references" support for arrow functions in VS Code back in the day.  Furious googling has failed to turn anything up so I think it's a thing of the past

There's a nice guide to the differences here: https://dmitripavlutin.com/differences-between-arrow-and-regular-functions/

My own preference is arrow functions within a module and full function if I export it.  I don't know that that is anything more than habit these days.

#### It's a `class` issue

One thing I noticed about your code that would definitely spark questions here and there with some is that you used `class`es.  It's worth knowing that classes only arrived in JavaScript with ES2015 and so are a relatively new construct.  And one that much of the community loathes.  Most of that is an overreaction in my opinion.

That said, I think classes are fine but only use them if it feels the obvious choice.  The [TypeScript compiler](https://github.com/microsoft/TypeScript) has no classes in at all.  Neither does [`ts-loader`](https://github.com/TypeStrong/ts-loader) save for test code (that may apply to the TypeScript compiler also).  You will rarely (if ever) see someone use inheritance extensively in a TypeScript codebase even if they do use classes.  It can be done but save for tutorials I've never seen it in the wild.  Nor would I probably want to!

Full disclosure: we use classes in ix for a number of use cases (historically that was the way to go to write React applications; though that has changed with React Hooks). 

#### TypeScript for functional programmers

I wonder if this TypeScript for Functional Programmers guide might be interesting to you?  The link below jumps straight to the Haskell similarities section:

https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes-func.html#concepts-similar-to-haskell

**Comment by johnnyreilly** [comment]
> The caller descends into parenthesis hell. I tried that style for a while - the syntax and tooling doesn't support it well enough and the code has a completely different flavour.

One final thought I wonder if the proposed pipeline operator that's working it's way into JavaScript might help you?

https://github.com/tc39/proposal-pipeline-operator

See also: 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Pipeline_operator
https://github.com/microsoft/TypeScript/pull/38305

**Comment by graeme-lockley** [comment]
"No Classes" - sounds like a line from the Incredibles.  That's great feedback as I am interested in understanding what idiomatic TypeScript looks like.  Looking through some code bases I did indeed notice no classes however there was an abundance of interfaces in order to assist with typing.

The "TS for Functional Programmers" has the following re currying and composition:

> Point-free programming
> 
> Point-free programming — heavy use of currying and function composition — is possible in JavaScript, but can be verbose. In TypeScript, type inference often fails for point-free programs, so you’ll end up specifying type parameters instead of value parameters. The result is so verbose that it’s usually better to avoid point-free programming.

This makes complete sense and is consistent with my own experience of parenthesis hell.

**Comment by graeme-lockley** [comment]
Just thinking... without classes one is then looking functional with side-effects - probably more like OCaml than Haskell.  That starts to introduce smells of monadic state because, to create a structure that is consistently manipulated through a collection of cooperating functions, one either needs global state or monadic state.

Actually more like just rambling rather than just thinking.

**Comment by johnnyreilly** [comment]
> Looking through some code bases I did indeed notice no classes however there was an abundance of interfaces in order to assist with typing.

TL;DR is `interface`s enforce type structure.

Longer explanation: so it's worth knowing that `interfaces` in TypeScriptland are nothing like `interfaces` in say C# / Java etc.  In a TypeScript context interfaces are used to model types.  Interestingly TypeScript has two ways to do this: `interface`s and `type`s - `interface`s have always been part of TypeScript; `type`s came later.  

```ts
// this is an interface
interface IsARectangle {
    height: number;
    width: number;
}

// this is the equivalent type 
type IsARectangle = {
    height: number;
    width: number;
}

// here is a an object literal which satisfies the type / interface
const isARectange: IsARectangle = {
    height: 20,
    width: 30
}
```


An open source friend of mine @basarat has a little video on the topic here:

https://www.youtube.com/watch?v=IXAT3If0pGI

**Comment by johnnyreilly** [comment]
> probably more like OCaml than Haskell

Fun fact: there's a competitor to TypeScript in the statically typed JavaScript space called "Flow". It's built by Facebook.  I'm not sure how much it's maintained these days, but one of the interesting things about it is that it was built with OCaml

https://opam.ocaml.org/packages/flow_parser/#:~:text=It%20produces%20an%20AST%20that,compiled%20to%20JavaScript%20using%20js_of_ocaml.

By contrast TypeScript is written in TypeScript (and for most of it's history has been - obviously the for first version that couldn't be true :smile:)
